validation/glodap_fit.jl

```diff
@@ -0,0 +1,143 @@
+using CSV, DataFrames, Random, CairoMakie
+using OceanBioME: CarbonChemistry
+
+using OceanBioME.Models: teos10_density, teos10_polynomial_approximation
+using OceanBioME.Models.CarbonChemistryModel: K0, K1, K2, KF
+
+
+
+# get the glodap data - downloaded "merged" file from https://www.ncei.noaa.gov/data/oceans/ncei/ocads/data/0283442/
+data = CSV.read("validation/GLODAP.csv", DataFrame)
+
+DIC_name = "G2tco2"
+Alk_name = "G2talk"
+T_name = "G2temperature"
+S_name = "G2salinity"
+P_name = "G2pressure"
+Si_name = "G2silicate"
+PO_name = "G2phosphate"
+pH_name = "G2phtsinsitutp"
+depth_name = "G2depth"
+cruise_name = "G2cruise"
+fCO2_name = "G2fco2"
+lon_name = "G2longitude"
+lat_name = "G2latitude"
+
+DIC_name, Alk_name, T_name, S_name, P_name, Si_name, PO_name, pH_name
+
+# remove low quality data and non QCed
+# and drop a weird point from cruise with a massive errors (only 2 points)
+#=data = filter(row -> any([row[name*"f"] == 2 for name in [DIC_name, Alk_name, S_name]]) && 
+                     !any([row[name] == -9999 for name in [DIC_name, Alk_name, T_name, S_name]]) &&
+                     #row[cruise_name] != 5005 &&
+                     #row[cruise_name] != 345 &&
+                     row[Alk_name] > 1500 &&
+                     row["G2phtsqc"] == 1,  data)
+=#
+# if we don't have Si and PO₄ the error gets a lot bigger but not in the surface values (as much)
+data = filter(row -> any([row[name*"f"] == 2 for name in [DIC_name, Alk_name, S_name, pH_name]]) && 
+                     !any([row[name] == -9999 for name in [DIC_name, Alk_name, T_name, S_name, P_name, Si_name, PO_name, pH_name]]) &&
+                     row["G2expocode"] != "325020210316" &&
+                     row["G2expocode"] != "33RO20071215" &&
+                     row[cruise_name] != 270 &&
+                     #row[Alk_name] > 1500 &&
+                     row["G2phtsqc"] == 1,
+                     data)
+
+Nrows = size(data, 1)
+
+# setup the model
+
+density_function = teos10_density
+
+# pH_error = 0.00022 ± 0.01223 -> mean is zero
+# pCO₂_error = -3 ± 28.8 -> mean is also zero
+# with teos polynomial appriximation error is almost identical in pCO2 but slightly bigger in pH
+carbon_chemistry = CarbonChemistry()
+
+# compare
+glodap_pH = zeros(Nrows)
+glodap_pCO₂ = zeros(Nrows)
+
+model_pH  = zeros(Nrows)
+model_pCO₂ = zeros(Nrows)
+
+pH_error  = zeros(Nrows)
+pCO₂_error  = zeros(Nrows)
+
+Threads.@threads for n in 1:Nrows
+    T = data[n, T_name]
+    S = data[n, S_name]
+    P = data[n, P_name] * 0.1 # dbar to bar
+
+    lon = data[n, lon_name]
+    lat = data[n, lat_name]
+
+    density = density_function(T, S, P, lon, lat)
+
+    DIC = data[n, DIC_name] * density * 1e-3
+    Alk = data[n, Alk_name] * density * 1e-3
+    
+    silicate = data[n, Si_name] * density * 1e-3
+    phosphate = data[n, PO_name] * density * 1e-3
+
+    silicate = ifelse(silicate == -9999, 0, silicate)
```

Comment:
the ifself statement should probably go before the units of silicate and phosphate are converted?